internal/storage/chunk_uploader_test.go

```diff
@@ -0,0 +1,330 @@
+	obj := t.obj
+	chunkSize := googleapi.MinUploadChunkSize
+	uploadSizeFactors := [5]float64{1.243, 1.32536, 1.0, 1.63464, 1.99}
+	var contents []string
+	var totalContentSize int
+	var numChunks int
+	for _, factor := range uploadSizeFactors {
+		b := make([]byte, int(factor*float64(chunkSize)))
+		_, err := rand.Read(b)
+		AssertEq(nil, err)
+		contents = append(contents, string(b))
+		contentSize := len(b)
+		totalContentSize += contentSize
+		numChunks += int(math.Ceil(float64(contentSize) / float64(chunkSize)))
+	}
+
+	var numCallbacks int
+	var numBytesUploaded int64
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, chunkSize,
+		func(n int64) {
+			numBytesUploaded = n
+			numCallbacks++
+		})
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	for _, content := range contents {
+		err = uploader.UploadChunkAsync(strings.NewReader(content))
+		AssertEq(nil, err)
+	}
+
+	AssertEq(Writing, uploader.state)
+
+	// time.Sleep(1000 * time.Millisecond)
+	if numCallbacks > 0 {
+		AssertGe(numChunks, numCallbacks)
+		AssertGe(totalContentSize, numBytesUploaded)
+	}
+	AssertEq(numBytesUploaded, uploader.BytesUploadedSoFar())
+	AssertEq(totalContentSize, uploader.bytesUploadInitiatedSoFar())
+}
+
+func (t *ChunkUploaderTest) TestCloseWithoutWrite() {
+	obj := t.obj
+
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, 1000, nil)
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	o, err := uploader.Close()
+	AssertNe(nil, o)
+	AssertEq(nil, err)
+	AssertEq(o.Name, "test_object")
+	AssertEq(0, o.Size)
+}
+
+func (t *ChunkUploaderTest) TestCloseWithSingleUpload() {
+	obj := t.obj
+	chunkSize := 1000
+	contentSize := len(t.content)
+
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, chunkSize, nil)
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	err = uploader.UploadChunkAsync(t.req.Contents)
+	AssertEq(nil, err)
+
+	o, err := uploader.Close()
+	AssertNe(nil, o)
+	AssertEq(nil, err)
+	AssertEq(o.Name, "test_object")
+	AssertEq(contentSize, o.Size)
+}
+
+func (t *ChunkUploaderTest) TestCloseWithMultipleSmallUploads() {
+	obj := t.obj
+	chunkSize := 1000
+	contents := [6]string{"hfue", "yrf8934h9", "iru328ry", "iy3rh34r489y8", "ie32hr83hr43rt9y8", "i9r38ry2u9j9"}
+	var contentSize int
+
+	uploader, err := NewChunkUploader(context.Background(), obj, t.req, chunkSize, nil)
+	AssertNe(nil, uploader)
+	AssertEq(nil, err)
+
+	for _, content := range contents {
+		err = uploader.UploadChunkAsync(strings.NewReader(content))
+		AssertEq(nil, err)
+		contentSize += len(content)
+	}
+
+	o, err := uploader.Close()
+	AssertNe(nil, o)
+	AssertEq(nil, err)
+	AssertEq(o.Name, "test_object")
+	AssertEq(contentSize, o.Size)
+}
+
+func (t *ChunkUploaderTest) TestCloseWithMultipleBigUploads() {
+	obj := t.obj
+	chunkSize := 4
+	contents := [6]string{"hfuerj3ifj3920",
+		"yrf8934h9or329",
+		"iru328ryo9dj320j3",
+		"iy3rh34r489y89j309",
+		"ie32hr83hr43rt9y8do93j9",
+		"i9r38ry2u9j9orj3"}
+	var contentSize int
```

Comment:
Negative test cases are missing
when empty contents are passed to uploadChunkAsync?
when io.EOF is returned
Can you trigger an error scenario - invalid io.Reader and what happens when you call close?

Reply:
> when empty contents are passed to uploadChunkAsync?

Added test for passing empty content to uploadChunkAsync.

> when io.EOF is returned

io.EOF is never returned, so removed that check from implementation.

> invalid io.Reader and what happens when you call close?

This needs work.

Reply:
Added test for failed uploader and another test for showing its sucessful closure.